### DISCUSSION TITLE
Engine Connection does not resolve server_version_info when pool is reassigned

### DISCUSSION THREAD

**Discussion by infohash** [sqlalchemy/sqlalchemy#11171] [Usage Questions] [answered]
# How To Reproduce

```python
>>> from sqlalchemy import create_engine
>>> engine = create_engine(url="postgresql+psycopg2://admin:admin@localhost:5432/db")
>>> engine.dialect.server_version_info is None
True
>>> with engine.begin() as conn:
...     conn.execute(text(...))
>>> engine.dialect.server_version_info
(16, 0)
```

But if you explicitly assign a different pool to the engine, [Connection](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection) does not resolve `server_version_info`.
```python
>>> from sqlalchemy import pool, create_pool_from_url
null_pool = create_pool_from_url(url="postgresql+psycopg2://admin:admin@localhost:5432/db", poolclass=pool.NullPool)

engine = create_engine(url="postgresql+psycopg2://admin:admin@localhost:5432/db")
engine.pool = null_pool

>>> with engine.begin() as conn:
...     conn.execute(text(...))
>>> engine.dialect.server_version_info is None
True
```

The problem with this is, [AutomapBase](https://docs.sqlalchemy.org/en/20/orm/extensions/automap.html#sqlalchemy.ext.automap.AutomapBase) throws an error if `server_version_info` is `None`:
```python
>>> from sqlalchemy.ext.automap import automap_base
>>> auto_map_base = automap_base()
>>> auto_map_base.prepare(autoload_with=engine, schema="public")
>>> auto_map_base.classes.items()
if self.server_version_info >= (12,)
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: '>=' not supported between instances of 'NoneType' and 'tuple'
```

This is because [Connection](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Connection) didn't resolve `server_version_info` so it is `None` by default.

This error does not occur if I don't explicitly assign a pool AFTER creating the `Engine` instance:
```python
>>> engine = create_engine(url="postgresql+psycopg2://admin:admin@localhost:5432/db")
>>> with engine.begin() as conn:
...     conn.execute(text(...))
>>> auto_map_base = automap_base()
>>> auto_map_base.prepare(autoload_with=engine, schema="public")
[('app', <class 'sqlalchemy.ext.automap.app'>), ('test', <class 'sqlalchemy.ext.automap.test'>)]
```

It's fine if you assign a different pool during the initialization of the engine:
```python
>>> engine = create_engine(url="postgresql+psycopg2://admin:admin@localhost:5432/db", poolclass=pool.NullPool)
>>> with engine.begin() as conn:
...     conn.execute(text(...))
>>> auto_map_base = automap_base()
>>> auto_map_base.prepare(autoload_with=engine, schema="public")
[('app', <class 'sqlalchemy.ext.automap.app'>), ('test', <class 'sqlalchemy.ext.automap.test'>)]
```

## Cause

```python
from sqlalchemy import pool, create_engine, create_pool_from_url
engine = create_engine(url="postgresql+psycopg2://admin:admin@localhost:5432/db")
null_pool = create_pool_from_url(url="postgresql+psycopg2://admin:admin@localhost:5432/db", poolclass=pool.NullPool)
engine.pool = null_pool
```

[Engine](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Engine)`.pool` is a property. When a new value is assigned to it, it triggers `setter` descriptor 👇:
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/engine/base.py#L3361-L3364

which executes this function 👇:
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/event/registry.py#L103-L118

See the line 110 👆 above, it is popping a `weakref` from `_collection_to_key`. `_collection_to_key` contains a dictionary of weak references to many objects but one such in particular which is causing the issue after being popped is 👇:
```python
>>> {<weakref at 0x000001B5979FCD60; to 'function' at 0x000001B5979FAF20 (on_connect)>: (1879445071184, 'connect', 1879444533024), <weakref at 0x000001B5979FCE00; to 'function' at 0x000001B5979FB880 (go)>: (1879445071184, 'connect', 1879444535264)}
```

Popping this 👆  will soon cause an issue when the connection will be created. See the output of the popped value of the weakref. One of the weak reference is pointing to the function `go`. This function will soon come to the picture. Here's what happens when the connection is created 👇:
```python
from sqlalchemy import create_engine, create_pool_from_url
engine = create_engine(url="postgresql+psycopg2://admin:admin@localhost:5432/db")
null_pool = create_pool_from_url(url="postgresql+psycopg2://admin:admin@localhost:5432/db", poolclass=pool.NullPool)
engine.pool = null_pool

>>> with engine.begin() as conn:
...     conn.execute(text(...))
```
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/engine/base.py#L3284
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/engine/base.py#L150
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/engine/base.py#L3308
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/pool/base.py#L449
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/pool/base.py#L1263
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/pool/base.py#L712
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/pool/impl.py#L177
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/pool/base.py#L390
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/pool/base.py#L674
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/pool/base.py#L910-L914
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/event/attr.py#L483
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/event/attr.py#L496-L498

Here 👆 is the problem. The weakref was popped earlier by the pool setter descriptor so `self.listeners` is empty 👇:
```python
>>> self.listeners
deque([])
```

## If it was not empty, the function `go` would be executed 👇:
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/util/langhelpers.py#L1902-L1915

See the line 1909 above 👆, it executes 👇:
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/engine/create.py#L741

This executes driver specific `initialize` method of **asyncpg** or **psycopg2** 👇:
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/dialects/postgresql/psycopg2.py#L676-L677
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/dialects/postgresql/base.py#L3066-L3067
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/engine/default.py#L521-L525
https://github.com/sqlalchemy/sqlalchemy/blob/414025a631904f4f3c767a055a8265b9cbb9407a/lib/sqlalchemy/dialects/postgresql/base.py#L3334-L3345

# Workaround
When a different pool is assigned after creating an engine, I also assign the value to `server_version_info` so that `AutoMapBase`  does not encounter comparison with `None` type 👇:

```python
null_pool = create_pool_from_url(url="postgresql+psycopg2://admin:admin@localhost:5432/db", poolclass=pool.NullPool)

engine = create_engine(url="postgresql+psycopg2://admin:admin@localhost:5432/db")
engine.pool = null_pool
engine.dialect.server_version_info = (16, 0)

with engine.begin() as conn:
    conn.execute(text(...))
auto_map_base = automap_base()

>>> auto_map_base.prepare(autoload_with=engine, schema="public")
[('app', <class 'sqlalchemy.ext.automap.app'>), ('test', <class 'sqlalchemy.ext.automap.test'>)]
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

You are not supported to assign a pool to an engine after creation. Pass the pool to create_engine https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.pool
